service/collector/ec2instances.go

```diff
@@ -0,0 +1,241 @@
+			}
+
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// Describe emits the description for the metrics collected here.
+func (e *EC2Instances) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- ec2InstanceStatus
+	return nil
+}
+
+// collectForAccount collects and emits our metric for one AWS account.
+//
+// We gather two separate collections first, then match them by instance ID:
+// - instance information, including tags, only for those tagged for our installation
+// - instance status information
+func (e *EC2Instances) collectForAccount(ch chan<- prometheus.Metric, awsClients clientaws.Clients) error {
+	account, err := e.helper.AWSAccountID(awsClients)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	// Collect instance status info.
+	// map key will be the instance ID.
+	instanceStatuses := map[string]*ec2.InstanceStatus{}
+	{
+		input := &ec2.DescribeInstanceStatusInput{
+			IncludeAllInstances: aws.Bool(true),
+			MaxResults:          aws.Int64(1000),
+		}
+
+		for {
+			o, err := awsClients.EC2.DescribeInstanceStatus(input)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			// collect statuses
+			for _, s := range o.InstanceStatuses {
+				instanceStatuses[*s.InstanceId] = s
+			}
+
+			if o.NextToken == nil {
+				break
+			}
+			input.SetNextToken(*o.NextToken)
+		}
+	}
+
+	// Collect instance info.
+	instances := map[string]*ec2.Instance{}
+	{
+		input := &ec2.DescribeInstancesInput{
+			Filters: []*ec2.Filter{
+				&ec2.Filter{
+					Name: aws.String(fmt.Sprintf("tag:%s", tagInstallation)),
+					Values: []*string{
+						aws.String(e.installationName),
+					},
+				},
+			},
+			MaxResults: aws.Int64(1000),
+		}
+
+		for {
+			o, err := awsClients.EC2.DescribeInstances(input)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			for _, reservation := range o.Reservations {
+				for _, instance := range reservation.Instances {
+					instances[*instance.InstanceId] = instance
+				}
+			}
+
+			if o.NextToken == nil {
+				break
+			}
+			input.SetNextToken(*o.NextToken)
+		}
+	}
+
+	// Iterate over found instances and emit metrics.
+	for instanceID := range instances {
+
+		// Skip if we don't have a status for this instance.
+		statuses, statusesAvailable := instanceStatuses[instanceID]
+		if !statusesAvailable {
+			continue
+		}
+
+		var cluster, installation, organization, state, status string
+		for _, tag := range instances[instanceID].Tags {
+			switch *tag.Key {
+			case tagCluster:
+				cluster = *tag.Value
+			case tagInstallation:
+				installation = *tag.Value
+			case tagOrganization:
+				organization = *tag.Value
+			}
+		}
+
+		up := 0
+		if statuses.InstanceState.Name != nil {
+			state = strings.ToLower(*statuses.InstanceState.Name)
+		}
+		if statuses.InstanceStatus.Status != nil {
+			status = strings.ToLower(*statuses.InstanceStatus.Status)
+		}
+		if state == "running" && status == "ok" {
```

Comment:
I don't get why we have these labels. Wouldn't this be the value of the metric?

Reply:
e.g:
i'd expect
```
aws_operator_ec2_status{account="270935918670",cluster_id="3tiea",ec2instance="i-03b17014ebbe7deb8",installation="ginger",organization="giantswarm"} 1
```
for a good boy
and
```
aws_operator_ec2_status{account="270935918670",cluster_id="3tiea",ec2instance="i-03b17014ebbe7deb8",installation="ginger",organization="giantswarm"} 0
```
for a bad boy (i'm assuming i'm missing a nuance here :D )

Reply:
As I said, the value is only 1 when state is "running" AND status is "ok". That's a healthy, operating instance.

An instance that has failed healthchecks will have a `status` value != `ok`. Likely the state string appearing will be `impaired`.